crates/turborepo-lib/src/task_hash.rs

```diff
@@ -115,14 +124,74 @@ impl PackageInputsHashes {
-                    Err(err) => return Some(Err(err.into())),
+                // Try hashing with the daemon, if we have a connection. If we don't, or if we
+                // timeout or get an error, fallback to local hashing
+                let hash_object = daemon
+                    .as_mut()
+                    .and_then(|daemon| {
+                        let handle = handle.clone();
+                        // We need an async block here because the timeout must be created with an
+                        // active tokio context. Constructing it directly in
+                        // the rayon thread doesn't provide one and will crash at runtime.
+                        handle
+                            .block_on(async {
+                                tokio::time::timeout(
+                                    Duration::from_millis(100),
+                                    daemon.get_file_hashes(package_path, &task_definition.inputs),
+                                )
+                                .await
+                            })
+                            .map_err(|e| {
+                                tracing::debug!(
+                                    "daemon file hashing timed out for {}",
+                                    package_path
+                                );
+                                e
+                            })
+                            .ok() // If we timed out, we don't need to error,
+                                  // just return None so we can move on to local
+                    })
+                    .and_then(|result| {
+                        match result {
+                            Ok(hashes_resp) => Some(
+                                hashes_resp
+                                    .file_hashes
+                                    .into_iter()
+                                    .map(|(path, hash)| {
+                                        (
+                                            RelativeUnixPathBuf::new(path)
+                                                .expect("daemon returns relative unix paths"),
+                                            hash,
+                                        )
+                                    })
+                                    .collect::<HashMap<_, _>>(),
+                            ),
+                            Err(e) => {
+                                // Daemon could've failed for various reasons. We can still try
+                                // local hashing.
+                                tracing::debug!(
+                                    "daemon file hashing failed for {}: {}",
+                                    package_path,
+                                    e
+                                );
+                                None
+                            }
+                        }
+                    });
+                let mut hash_object = match hash_object {
```

Comment:
not very important but you could probably use `or_else` or `unwrap_or_else` here.

Reply:
I looked at doing that, but it gets awkward since we're starting with an `Option`, but the fallback calculation produces a `Result` where we want to short-circuit, but the overall closure returns an `Option<Result<_, _>>`. `?` doesn't really work in that context, so you end up with the `match` either way.